### ISSUE TITLE
Per-document progress notifications on replications

### ISSUE THREAD

**Issue by snej** [couchbase/couchbase-lite-ios#293]
It would be useful for replications to notify which documents have been pushed/pulled, or if documents failed, what the errors were.

**Comment by snej** [comment]
FYI, the upcoming new replicator supports this via NSProgress objects, but it's not accessible through the public API.

**Comment by tleyden** [comment]
What about allowing api users to have a registered callback function whose result would help the replicator decide what to do?  

```
push.addErrorCallback(
  func(docId, error) -> what_to_do {
    if docId == "settings" {
       // nothing will work if the settings doc doesn't sync, so just give up at this point
       return abort_replication
    }
    // as long as it's not the settings doc, we can probably safely ignore it .. write some logs and/or alert the user, but keep going
    return keep_going!
  }
)
```

If a developer didn't register a callback function, then it would just keep the current behavior and stop -- this does have the advantage of having errors of being very easy to notice since they bring the whole replication to a screeching halt.

**Comment by snej** [comment]
This has [come up as a problem, on the forum](https://forums.couchbase.com/t/question-about-conflicts/7211/7).

**Comment by snej** [comment]
Very good point, from the above forum thread, about the problem of a pushed doc being rejected on the server due to a permission change:

>Consider a chat room, where owners can change the name of the chat room. So imagine owner Bob attempts to make a name change, first checking that they are an owner, which they are. Bob submits the name change, while at the same time Jim removes Bob from the list of owners. If Jim's change makes it to the server first then Bob's change fails a requireUser() check on the owners and gets rejected. In this case both clients did their job correctly but the error still happens. And now Bob has an inconsistent document, and it will be inconsistent indefinitely, with no indication at all that there is a problem. He has a different view of the world to everyone else.

>I guess my issue is that there's no work around - from what you're saying Bob has to wait for someone else to make a change at which point the document should be visible in the list of conflicts. And the change may never happen.

>I think marking the document in some way, (as being rejected maybe) would be enough that the client could take action to remove the failed version.

**Comment by BharathMG** [comment]
@snej  Hi, any updates on this? I'm facing the same issue in Android. Is forking Couchbase lite repo for changing this behaviour the only solution?

**Comment by pasin** [comment]
We have implemented this in iridium (2.5.0)
